app/handlers/base.py

```diff
@@ -20,6 +24,51 @@
 env, cfg = load_env()
 
 
+# Monkey-patch Python Social Auth's base handler
+#
+# See
+# https://github.com/python-social-auth/social-app-tornado/blob/master/social_tornado/handlers.py
+# for the original
+#
+# Python Social Auth documentation:
+# https://python-social-auth.readthedocs.io/en/latest/backends/implementation.html#auth-apis
+
+class PSABaseHandler(RequestHandler):
+    """
+    Mixin used by Python Social Auth
+    """
+    def user_id(self):
+        return self.get_secure_cookie('user_id')
+
+    def get_current_user(self):
+        user_id = self.user_id()
+        oauth_uid = self.get_secure_cookie('user_oauth_uid')
+        if user_id and oauth_uid:
+            user = User.query.get(int(user_id))
+            sa = user.social_auth.first()
+            if (sa.uid.encode('utf-8') == oauth_uid):
+                return user
+
+    def login_user(self, user):
+        sa = user.social_auth.first()
```

Comment:
What if there's no match? Where do new users get created?

Reply:
If there's no match, then the user does not get logged in, and then the pipeline proceeds the same way it usually does.